dask/array/tests/test_atop.py

```diff
@@ -290,6 +290,34 @@ def f(x):
     assert_eq(c, np.ones((5, 7)))
 
 
+def test_blockwise_new_axes_chunked():
+    def f(x):
+        return x[None, :] * 2
+
+    x = da.arange(0, 6, 1, chunks=2, dtype=np.int32)
+    y = da.blockwise(f, 'qa', x, 'a', new_axes={'q': (1, 1)}, dtype=x.dtype)
+    assert y.chunks == ((1, 1), (2, 2, 2))
+    assert_eq(y, np.array([[0, 2, 4, 6, 8, 10], [0, 2, 4, 6, 8, 10]], np.int32))
+
+
+def test_blockwise_no_args():
+    def f():
+        return np.ones((2, 3), np.float32)
+
+    x = da.blockwise(f, 'ab', new_axes={'a': 2, 'b': (3, 3)}, dtype=np.float32)
+    assert x.chunks == ((2,), (3, 3))
+    assert_eq(x, np.ones((2, 6), np.float32))
+
+
+def test_blockwise_no_array_args():
+    def f(dtype):
+        return np.ones((2, 3), dtype)
+
+    x = da.blockwise(f, 'ab', np.float32, None, new_axes={'a': 2, 'b': (3, 3)}, dtype=np.float32)
+    assert x.chunks == ((2,), (3, 3))
+    assert_eq(x, np.ones((2, 6), np.float32))
```

Comment:
It's quite cool to see these.  In the future I wonder if we might be able to simplify operations like `da.ones` and `da.random` to use blockwise.

Reply:
Quite possibly.

One thing I've found slightly annoying with this interface is that there isn't a field that directly tells you the shape you need to output; you have to compute it by taking the difference of the lower and upper corners from `array-location`. Do you think it's worth adding another field to the dict for this? Another candidate would be the `dtype`, which would make it easier to write generic callback functions without having to pass the dtype both to the callback function and to `map_blocks` itself (although I'm not sure how dtype inference would work).